### PR TITLE
Integration-specific settings (including Trace Analytics)

### PR DESCRIPTION
**Pull request by lucaspimentel**
Add integration-specific settings to `TracerSettings`, including "Trace Search and Analytics".

- configuration via variables
  - `DD_TRACE_ANALYTICS_ENABLED`
  - `DD_<integration>_ENABLED`
  - `DD_<integration>_ANALYTICS_ENABLED`
  - `DD_<integration>_ANALYTICS_SAMPLE_RATE`
- configuration via code
  - `Tracer.Settings.AnalyticsEnabled`
  - `Tracer.Settings.Integrations["<integration>"].Enabled`
  - `Tracer.Settings.Integrations["<integration>"].AnalyticsEnabled`
  - `Tracer.Settings.Integrations["<integration>"].AnalyticsSampleRate`

TODO
- add .NET tab to [the documentation](https://docs.datadoghq.com/tracing/trace_search_and_analytics#automatic-configuration)

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs [hunk 1]
```diff
@@ -58,10 +59,10 @@ public AspNetCoreMvc2Integration(object actionDescriptor, object httpContext)
                 }
 
                 string url = GetDisplayUrl(request).ToLowerInvariant();
-
                 SpanContext propagatedContext = null;
+                var tracer = Tracer.Instance;
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs]: Why the change to a local variable? verbosity?

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs]: Partly as (probably premature) optimization, since it removes several calls to the property getter (and this code is called by the profiler, which disables method inlining).

Partly because `Tracer.Instance` can change, and we should use the same `Tracer` instance throughout this code, even if `Tracer.Instance` changes halfway through.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs [hunk 2]
```diff
@@ -123,6 +129,12 @@ public AspNetCoreMvc2Integration(object actionDescriptor, object httpContext)
         {
             AspNetCoreMvc2Integration integration = null;
 
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName))
```
Comment (colin-higgins) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs]: In lieu of your comment on the tracer instance being able to change, does that matter here? I assume the settings are the same regardless?

File: src/Datadog.Trace/Configuration/TracerSettings.cs [hunk 3]
```diff
@@ -152,5 +182,23 @@ public static CompositeConfigurationSource CreateDefaultConfigurationSource()
...
+
+        internal double? GetIntegrationAnalyticsSampleRate(string name, bool enabledWithGlobalSetting)
+        {
+            var integrationSettings = Integrations[name];
+            var analyticsEnabled = integrationSettings.AnalyticsEnabled ?? (enabledWithGlobalSetting && AnalyticsEnabled);
+            var sampleRate = integrationSettings.AnalyticsSampleRate ?? AnalyticsSampleRate;
+
+            return analyticsEnabled
+                       ? sampleRate
+                       : (double?)null;
```
Comment (brettlangdon) [inline src/Datadog.Trace/Configuration/TracerSettings.cs]: interesting syntax... I get it... just calling out it is interesting. (re: `(double?)null`)

File: src/Datadog.Trace/Span.cs [hunk 4]
```diff
@@ -172,6 +173,34 @@ public Span SetTag(string key, string value)
...
+                    bool? boolean = value.ToBoolean();
+
+                    if (boolean == true)
+                    {
+                        // always sample
+                        SetMetric(Trace.Tags.Analytics, 1.0);
+                    }
+                    else if (boolean == false)
+                    {
+                        // never sample, remove the metric
```
Comment (brettlangdon) [inline src/Datadog.Trace/Span.cs]: ah, so this is interesting, since setting `0` is different than not setting at all.

If the metric is not set, then we inherit the sample rate from the agent.

If the metric is set to `0` then we override any agent sample rate to be `0` (don't sample).

In this case we actually do want to set to `0` instead of `null`. The idea is that if they explicitly call `span.SetTag(Trace.Tags.Analytics, false)` they are telling us to explicitly disable analytics for that span, which would mean setting as `0` instead of `null`.

Make sense?

File: src/Datadog.Trace/Configuration/TracerSettings.cs (hunk 3 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Configuration/TracerSettings.cs]: TL;DR: the code won't compile without the cast.

In the expression `a ? b : c`, `b` and `c` need to be of the same type or have an implicit conversion between them. In this case, we have a `double` and `null`. There is no implicit conversion between them, but both can be converted implicitly to `double?`. The compiler just needs help figuring this out. As long as one of the values is `double?` the other can be either `double` or `null`.

Casting `sampleRate` would have worked just as well:
```csharp
// cast sampleRate instead of null
return analyticsEnabled
           ? (double?)sampleRate
           : null;
```

File: src/Datadog.Trace/Span.cs (hunk 4 above)
Comment (lucaspimentel) [reply src/Datadog.Trace/Span.cs]: Good catch, thanks! Fixed in 8e5b3bb9ff8e5e6c1a72cc04385e8d1fdd19840a.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs (hunk 2 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetCoreMvc2Integration.cs]: Yes, it matters here because each `Tracer` instance can have different settings, but it's a bigger change here than simply adding a local variable. We'll revisit this one when we update this integration to follow the same pattern as the others.

To be clear, this is a very unlikely edge case where the user replaces the global `Tracer` while automatic instrumentation is happening (instead of setting it only once during application initialization).
